ecosystem/sep-direct-payment.md

```diff
@@ -0,0 +1,318 @@
+
+Each `fields` sub-object contains a key for each field name and an object with the following fields as the value:
+
+* `description`: description of field to show to user.
+* `choices`: list of possible values for the field.
+
+### Send
+
+#### Request
+
+```
+POST DIRECT_PAYMENT_SERVER/send
+Content-Type: application/json
+
+{
+  "amount": 100,
+  "require_receiver_info": ["address", "tax_id"],
+  "fields": {
+    "sender": {
+      "first_name": "Alice",
+      "last_name": "Jones"
+    },
+    "receiver": {
+      "first_name": "Bob",
+      "last_name": "Dillon",
+      "email_address": "bdillon@something.com"
+    },
+    "transaction": {
+      "routing_number": "442928834",
+      "account_number": "0029483242",
+      "type": "SEPA"
+    }
+  }
+}
+```
+
+This post requests attempts to initiate a payment through this anchor.  It should provide the amount and all the required fields (specified in the [`/info`](#info) endpoint).  The sending anchor can also provide a `require_receiver_info` object if it needs information on the receiver client that the receiving anchor can provide.
+
+If the request describes a valid transaction that this anchor can fulfill, we return a success response with details on what to send. If the request is not valid, or we need more info, we can return with an error response and expect the sending anchor to try again with updated values.
+
+##### Request Parameters
+
+Name | Type | Description
+-----|-----|------
+`require_receiver_info` | array | A list of [SEP-9](sep-0009.md) values requested for the receiving client
+`amount` | number | Amount of payment in destination currency
+`fields` | object | A key-pair object containing the values requested by the receiving anchor in their `/info` endpoint
+`lang` | string | (optional) Defaults to `en`. Language code specified using [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1).  Any human readable error codes or field descriptions will be returned in this language.
+
+#### Responses
+
+##### Success (200 ok)
+This is the successful case where a receiving anchor confirms that they can fulfill this payment as described. The response body should be a JSON object with the following values
+
+Name | Type | Description
+-----|------|------------
+`id` | string | Persistent identifier to check the status of this payment
+`stellar_account_id` | string | Stellar account to send payment to
+`stellar_memo_type` | string | Type of memo to attach to the stellar payment `(text | hash | id)`
+`stellar_memo` | string | The memo to attach to the stellar payment
+`receiver_info` | object | Key-value pairs of the information the sender requested via `require_receiver_info`
+
+##### Customer Info Needed (403 forbidden)
```

Comment:
I don't think 403 is appropriate here, 400 seems like the natural fit.

Reply:
i was actually really confused by what i thought 403 meant.  
http://w3.org/Protocols/rfc2616/rfc2616%20-sec10.html
"The server understood the request, but is refusing to fulfill it. Authorization will not help and the request SHOULD NOT be repeated"

My understanding was that 403 meant 'needs authorization' but now this says its not.  

400 says 
"The request could not be understood by the server due to malformed syntax"
which doesn't make sense here for sure

Reply:
That link gave me a 404, but check out the [spec](https://tools.ietf.org/html/rfc7231#section-6.5.1)

I would consider missing arguments a malformed request.

Reply:
I agree that 400 makes a bit more sense here (it's also a kitchen sink for any client side error): 
>   The 400 (Bad Request) status code indicates that the server cannot or
   **will not process the request due to something that is perceived to be
   a client error** (e.g., malformed request syntax, invalid request
   message framing, or deceptive request routing).

Reply:
According to the RFCs I think 403 works, but practically it will be confusing because the first thing many think of when we see that error code is credentials, even if the specs don't require it to be related to credentials. Note, there's also two definitions of 403 in two different RFCs, which I think contributes to the confusion around when to use it. [RFC7231](https://tools.ietf.org/html/rfc7231#section-6.5.3) obsoleted [RFC2616](https://tools.ietf.org/html/rfc2616#section-10.4.4).

We could also consider 422. 400 was defined to mean the request was incorrect in and of itself (e.g. format, wrong field, missing field, wrong type in a field, etc), and 422 [was added for WebDAV](https://tools.ietf.org/html/rfc4918#section-11.2) and is referenced in at least one other [non-WebDAV RFC](https://tools.ietf.org/html/rfc5789#section-2.2) and indicates there is something semantically incorrect with sending that request at this time. I think this more accurately reflects the issue here since the request parses correctly.

I think it's easy to argue that the request is bad because it's invalid given the state of the account since it is missing information, but the downside with using a 400 is a developer might jump to "what did I put wrong in this request" rather than "what about the state of the account is incomplete."

Stripe uses 402. GitHub uses 422.

I think 400 is ok. There might be some benefits to distinguishing the case with 422, but that might just replace one type of confusion with a different kind of confusion so I don't think it is a clear win.

Reply:
I don't think missing key-value pairs in the `"fields"` object is a valid use of 422, since the request is syntactically incorrect, not semantically incorrect. 

A good example for when 422 would make a lot of sense is where SEP-6 currently uses [403 Customer Info Needed](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0006.md#2-customer-information-needed-non-interactive). This case differs from the one we're discussing (even though they share the same name) because in SEP-6 the `"fields"` returned in the 403 response need to be sent to a _different endpoint_, namely [SEP-12's `/customer`](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0012.md#customer-put). 422 should be used in this case because the SEP-6 request was valid but couldn't be processed without a prior request to `/customer`, i.e. the semantics are incorrect.

Reply:
Moved to 400